File: java/agentbay/src/main/java/com/aliyun/agentbay/command/Command.java
package com.aliyun.agentbay.command;

import com.aliyun.agentbay.model.CommandResult;
import com.aliyun.agentbay.model.OperationResult;
import com.aliyun.agentbay.service.BaseService;
import com.aliyun.agentbay.session.Session;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.HashMap;
import java.util.Map;

/**
 * Async command execution service for session shells in the AgentBay cloud environment.
 * 
 * <p>Use this class for non-blocking command execution; for blocking/synchronous usage,
 * refer to the Command service in the sync API.
 */
public class Command extends BaseService {
    private static final String SERVER_SHELL = "wuying_shell";

    public Command(Session session) {
        super(session);
    }

    /**
     * Execute a shell command with default timeout (50000ms).
     *
     * @param command The shell command to execute
     * @return CommandResult Result object containing execution details
     */
    public CommandResult execute(String command) {
        return executeCommand(command, 50000);
    }

    /**
     * Execute a shell command with default timeout (50000ms).
     * Note: The input parameter is currently not used in the implementation.
     *
     * @param command The shell command to execute
     * @param input Input parameter (currently unused)
     * @return CommandResult Result object containing execution details
     */
    public CommandResult execute(String command, String input) {
        return executeCommand(command, 50000);
    }

    /**
     * Execute a shell command with specified timeout.
     *
     * @param command The shell command to execute
     * @param timeoutMs Timeout in milliseconds
     * @return CommandResult Result object containing execution details
     */
    public CommandResult executeCommand(String command, int timeoutMs) {
        return executeCommand(command, timeoutMs, null, null);
    }

    /**
     * Execute a shell command with optional working directory and environment variables.
     *
     * <p>Executes a shell command in the session environment with configurable timeout,
     * working directory, and environment variables. The command runs with session
     * user permissions in a Linux shell environment.
     *
     * @param command The shell command to execute
     * @param timeoutMs Timeout in milliseconds (default: 50000ms/50s).
     * @param cwd The working directory for command execution. If not specified,
     *            the command runs in the default session directory
     * @param envs Environment variables as a map of key-value pairs.
     *             These variables are set for the command execution only
     * @return CommandResult Result object containing:
     *         <ul>
     *           <li>success: Whether the command executed successfully (exit_code == 0)</li>
     *           <li>output: Command output for backward compatibility (stdout + stderr)</li>
     *           <li>exitCode: The exit code of the command execution (0 for success)</li>
     *           <li>stdout: Standard output from the command execution</li>
     *           <li>stderr: Standard error from the command execution</li>
     *           <li>traceId: Trace ID for error tracking (only present when exit_code != 0)</li>
     *           <li>requestId: Unique identifier for this API request</li>
     *           <li>errorMessage: Error description if execution failed</li>
     *         </ul>
     * @throws IllegalArgumentException If environment variables contain non-string keys or values
     *
     */
    public CommandResult executeCommand(String command, int timeoutMs, String cwd, Map<String, String> envs) {
        if (envs != null) {
            java.util.List<String> invalidVars = new java.util.ArrayList<>();
            for (Map.Entry<String, String> entry : envs.entrySet()) {
                Object key = entry.getKey();
                Object value = entry.getValue();

                if (!(key instanceof String)) {
                    invalidVars.add(String.format("key '%s' (type: %s)", key, key.getClass().getSimpleName()));
                }
                if (!(value instanceof String)) {
                    invalidVars.add(String.format("value for key '%s' (type: %s)", key, value.getClass().getSimpleName()));
                }
            }

            if (!invalidVars.isEmpty()) {
                throw new IllegalArgumentException(
                    "Invalid environment variables: all keys and values must be strings. " +
                    "Found invalid entries: " + String.join(", ", invalidVars)
                );
            }
        }

        try {
            Map<String, Object> args = new HashMap<>();
            args.put("command", command);
            args.put("timeout_ms", timeoutMs);

            if (cwd != null) {
                args.put("cwd", cwd);
            }

            if (envs != null) {
                args.put("envs", envs);
            }

            OperationResult result = callMcpTool("shell", args);

            if (result.isSuccess()) {
                String rawData = result.getData() == null ? "" : result.getData();
                ParsedShellPayload parsed = parseShellPayload(rawData);
                return new CommandResult(result.getRequestId(),
                        parsed.exitCode == 0,
                        parsed.stdout + parsed.stderr,
                        "",
                        parsed.exitCode,
                        parsed.stdout,
                        parsed.stderr,
                        parsed.traceId);
            } else {
                String rawErr = result.getErrorMessage() == null ? "" : result.getErrorMessage();
                ParsedShellPayload parsed = parseShellPayload(rawErr);
                String errorMessage = !parsed.stderr.isEmpty() ? parsed.stderr
                        : (rawErr.isEmpty() ? "Failed to execute command" : rawErr);
                int exitCode = parsed.exitCode != 0 ? parsed.exitCode : 1;
                return new CommandResult(result.getRequestId(), false,
                        parsed.stdout + parsed.stderr, errorMessage, exitCode,
                        parsed.stdout, parsed.stderr, parsed.traceId);
            }

        } catch (Exception e) {
            return new CommandResult("", false, "", "Failed to execute command: " + e.getMessage(), 1,
                    "", "", "");
        }
    }

    /**
     * Parsed payload from a shell-tool MCP response.
     *
     * <p>The MCP server's shell tool can return data in different shapes depending on the
     * sandbox image. The two known shapes are:
     * <ol>
     *   <li><b>Wrapped</b>: {@code {"exit_code":0,"stdout":"...","stderr":"...","traceId":"..."}}
     *       — used by code_latest and similar images.</li>
     *   <li><b>Raw</b>: the command's stdout returned verbatim with no JSON wrapping
     *       — used by openclaw / imgc-* and similar custom images.</li>
     * </ol>
     *
     * <p>This method detects the wrapped shape conservatively (requires both an integer
     * {@code exit_code} field <b>and</b> at least one of {@code stdout}/{@code stderr}, OR
     * both {@code stdout} and {@code stderr} together). Anything else — including JSON
     * objects/arrays/scalars and plain text — is treated as raw command output.
     */
    private static class ParsedShellPayload {
        final String stdout;
        final String stderr;
        final int exitCode;
        final String traceId;

        ParsedShellPayload(String stdout, String stderr, int exitCode, String traceId) {
            this.stdout = stdout;
            this.stderr = stderr;
            this.exitCode = exitCode;
            this.traceId = traceId;
        }
    }

    private static ParsedShellPayload parseShellPayload(String rawData) {
        if (rawData == null || rawData.isEmpty()) {
            return new ParsedShellPayload("", "", 0, "");
        }
        try {
            ObjectMapper mapper = new ObjectMapper();
            JsonNode node = mapper.readTree(rawData);
            if (looksLikeWrappedShellPayload(node)) {
                String stdout = node.has("stdout") ? node.get("stdout").asText() : "";
                String stderr = node.has("stderr") ? node.get("stderr").asText() : "";
                int exitCode = node.has("exit_code") ? node.get("exit_code").asInt() : 0;
                String traceId = node.has("traceId") ? node.get("traceId").asText() : "";
                return new ParsedShellPayload(stdout, stderr, exitCode, traceId);
            }
        } catch (Exception ignored) {
            // Not valid JSON — fall through to raw treatment.
        }
        // Raw stdout: return the entire payload verbatim.
        return new ParsedShellPayload(rawData, "", 0, "");
    }

    /**
     * Conservatively determines whether a JSON node looks like the shell-tool wrapped
     * envelope. We require strong signals to avoid mistaking a user command's JSON output
     * for the wrapper.
     */
    private static boolean looksLikeWrappedShellPayload(JsonNode node) {
        if (node == null || !node.isObject()) {
            return false;
        }
        boolean hasIntExitCode = node.has("exit_code") && node.get("exit_code").isInt();
        boolean hasStdout = node.has("stdout") && node.get("stdout").isTextual();
        boolean hasStderr = node.has("stderr") && node.get("stderr").isTextual();
        // Strong signal #1: integer exit_code + at least one textual stdout/stderr.
        if (hasIntExitCode && (hasStdout || hasStderr)) {
            return true;
        }
        // Strong signal #2: both textual stdout and stderr present (exit_code may default to 0).
        return hasStdout && hasStderr;
    }

    /**
     * Alias of executeCommand() for better ergonomics and LLM friendliness.
     *
     * @param command The shell command to execute
     * @param timeoutMs Timeout in milliseconds
     * @return CommandResult Result object containing execution details
     * @see #executeCommand(String, int, String, Map)
     */
    public CommandResult run(String command, int timeoutMs) {
        return executeCommand(command, timeoutMs);
    }

    /**
     * Alias of executeCommand() for better ergonomics and LLM friendliness.
     *
     * @param command The shell command to execute
     * @param timeoutMs Timeout in milliseconds
     * @param cwd The working directory for command execution
     * @param envs Environment variables as a map of key-value pairs
     * @return CommandResult Result object containing execution details
     */
    public CommandResult run(String command, int timeoutMs, String cwd, Map<String, String> envs) {
        return executeCommand(command, timeoutMs, cwd, envs);
    }

    /**
     * Alias of executeCommand() for better ergonomics and LLM friendliness.
     *
     * @param command The shell command to execute
     * @param timeoutMs Timeout in milliseconds
     * @return CommandResult Result object containing execution details
     */
    public CommandResult exec(String command, int timeoutMs) {
        return executeCommand(command, timeoutMs);
    }

    /**
     * Alias of executeCommand() for better ergonomics and LLM friendliness.
     *
     * @param command The shell command to execute
     * @param timeoutMs Timeout in milliseconds
     * @param cwd The working directory for command execution
     * @param envs Environment variables as a map of key-value pairs
     * @return CommandResult Result object containing execution details
     */
    public CommandResult exec(String command, int timeoutMs, String cwd, Map<String, String> envs) {
        return executeCommand(command, timeoutMs, cwd, envs);
    }

}

File: golang/pkg/agentbay/command/command.go
package command

import (
	"encoding/json"
	"fmt"

	mcp "github.com/aliyun/wuying-agentbay-sdk/golang/api/client"
	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay/models"
)

// CommandResult represents the result of a command execution
type CommandResult struct {
	// Embed the basic API response structure
	models.ApiResponse
	// Success indicates whether the command execution was successful
	Success bool `json:"success"`
	// Output contains the command execution output (for backward compatibility, equals stdout + stderr)
	Output string `json:"output"`
	// ErrorMessage contains error message if the operation failed
	ErrorMessage string `json:"error_message,omitempty"`
	// ExitCode is the exit code of the command execution. Default is 0.
	ExitCode int `json:"exit_code"`
	// Stdout is the standard output from the command execution
	Stdout string `json:"stdout"`
	// Stderr is the standard error from the command execution
	Stderr string `json:"stderr"`
	// TraceID is the trace ID for error tracking. Only present when exit_code != 0. Used for quick problem localization.
	TraceID string `json:"trace_id,omitempty"`
}

// Command handles command execution operations in the AgentBay cloud environment.
type Command struct {
	Session interface {
		GetAPIKey() string
		GetClient() *mcp.Client
		GetSessionId() string
		CallMcpTool(toolName string, args interface{}) (*models.McpToolResult, error)
	}
}

// NewCommand creates a new Command instance
func NewCommand(session interface {
	GetAPIKey() string
	GetClient() *mcp.Client
	GetSessionId() string
	CallMcpTool(toolName string, args interface{}) (*models.McpToolResult, error)
}) *Command {
	return &Command{
		Session: session,
	}
}

// ==================== Option Definitions ====================

// commandOptions holds the configuration for command execution
type commandOptions struct {
	timeoutMs int
	cwd       string
	envs      map[string]string
}

// CommandOption is a function type for configuring ExecuteCommand options.
// This enables the Functional Options pattern for flexible and extensible API design.
type CommandOption func(*commandOptions)

// WithTimeoutMs sets the timeout for command execution in milliseconds.
//
// Example:
//
//	cmd.ExecuteCommand("ls -la", WithTimeoutMs(5000))
func WithTimeoutMs(timeoutMs int) CommandOption {
	return func(opts *commandOptions) {
		opts.timeoutMs = timeoutMs
	}
}

// WithCwd sets the working directory for command execution.
// If not set, the command runs in the default session directory.
//
// Example:
//
//	cmd.ExecuteCommand("pwd", WithCwd("/tmp"))
func WithCwd(cwd string) CommandOption {
	return func(opts *commandOptions) {
		opts.cwd = cwd
	}
}

// WithEnvs sets environment variables for command execution.
// These variables are set for the command execution only.
//
// Example:
//
//	cmd.ExecuteCommand("echo $VAR", WithEnvs(map[string]string{"VAR": "value"}))
func WithEnvs(envs map[string]string) CommandOption {
	return func(opts *commandOptions) {
		opts.envs = envs
	}
}

// ExecuteCommand executes a shell command in the session environment.
//
// This method supports both the legacy signature (command string, timeoutMs ...int)
// and the Functional Options pattern for flexible configuration.
//
// Legacy usage (backward compatible):
//   - cmd.ExecuteCommand("ls -la")
//   - cmd.ExecuteCommand("ls -la", 5000)
//
// Functional Options usage (recommended for new code):
//   - cmd.ExecuteCommand("ls -la", WithTimeoutMs(5000))
//   - cmd.ExecuteCommand("pwd", WithCwd("/tmp"), WithEnvs(map[string]string{"VAR": "value"}))
//
// Parameters:
//   - command: The shell command to execute
//   - options: Either an int (timeoutMs in milliseconds) for legacy usage, or
//     CommandOption functions for Functional Options pattern.
//     Default is 50000ms (50s).
//
// Returns:
//   - *CommandResult: Result containing command output, exit code, stdout,
//     stderr, trace_id, and request ID
//   - error: Error if the operation fails
//
// Example:
//
//	// Default usage
//	cmd.ExecuteCommand("ls")
//
//	// Legacy usage (backward compatible)
//	cmd.ExecuteCommand("ls", 5000)
//
//	// New style with Functional Options
//	cmd.ExecuteCommand("ls", WithTimeoutMs(5000))
//
//	// Combined options
//	cmd.ExecuteCommand("pwd",
//	    WithTimeoutMs(5000),
//	    WithCwd("/tmp"),
//	    WithEnvs(map[string]string{"FOO": "bar"}),
//	)
func (c *Command) ExecuteCommand(command string, options ...interface{}) (*CommandResult, error) {
	// Default configuration
	opts := &commandOptions{
		timeoutMs: 1000, // Default 1 second
		cwd:       "",
		envs:      nil,
	}

	// Handle both old and new styles
	for _, opt := range options {
		switch v := opt.(type) {
		case int: // Old: ExecuteCommand("ls", 5000)
			if v > 0 {
				opts.timeoutMs = v
			}
		case CommandOption: // New: ExecuteCommand("ls", WithTimeoutMs(5000))
			if v != nil {
				v(opts)
			}
		}
	}

	return c.executeCommandInternal(command, opts.timeoutMs, opts.cwd, opts.envs)
}

// Run is an alias of ExecuteCommand.
func (c *Command) Run(command string, options ...interface{}) (*CommandResult, error) {
	return c.ExecuteCommand(command, options...)
}

// Exec is an alias of ExecuteCommand.
func (c *Command) Exec(command string, options ...interface{}) (*CommandResult, error) {
	return c.ExecuteCommand(command, options...)
}

// executeCommandInternal is the internal implementation of command execution
func (c *Command) executeCommandInternal(
	command string,
	timeout int,
	cwd string,
	envs map[string]string,
) (*CommandResult, error) {
	// Note: In Go, envs is typed as map[string]string, which enforces type safety at compile time.
	// All keys and values are guaranteed to be strings by the type system.
	// This validation is kept for consistency with other SDKs and documentation purposes.

	// Build request arguments
	args := map[string]interface{}{
		"command":    command,
		"timeout_ms": timeout,
	}
	if cwd != "" {
		args["cwd"] = cwd
	}
	if len(envs) > 0 {
		args["envs"] = envs
	}

	// Use Session's CallMcpTool method
	result, err := c.Session.CallMcpTool("shell", args)
	if err != nil {
		return nil, fmt.Errorf("failed to execute command: %w", err)
	}

	if result.Success {
		parsed := parseShellPayload(result.Data)
		return &CommandResult{
			ApiResponse: models.ApiResponse{
				RequestID: result.RequestID,
			},
			Success:      parsed.exitCode == 0,
			Output:       parsed.stdout + parsed.stderr,
			ExitCode:     parsed.exitCode,
			Stdout:       parsed.stdout,
			Stderr:       parsed.stderr,
			TraceID:      parsed.traceID,
			ErrorMessage: result.ErrorMessage,
		}, nil
	} else {
		rawErr := result.ErrorMessage
		parsed := parseShellPayload(rawErr)
		effectiveExitCode := parsed.exitCode
		if effectiveExitCode == 0 {
			effectiveExitCode = 1
		}
		effectiveError := parsed.stderr
		if effectiveError == "" {
			effectiveError = rawErr
		}
		if effectiveError == "" {
			effectiveError = "Failed to execute command"
		}
		return &CommandResult{
			ApiResponse: models.ApiResponse{
				RequestID: result.RequestID,
			},
			Success:      false,
			Output:       parsed.stdout + parsed.stderr,
			ExitCode:     effectiveExitCode,
			Stdout:       parsed.stdout,
			Stderr:       parsed.stderr,
			TraceID:      parsed.traceID,
			ErrorMessage: effectiveError,
		}, nil
	}
}

// parsedShellPayload holds the parsed fields from a shell-tool MCP response.
type parsedShellPayload struct {
	stdout   string
	stderr   string
	exitCode int
	traceID  string
}

// looksLikeWrappedShellPayload conservatively determines whether a parsed JSON
// map looks like the shell-tool wrapped envelope {exit_code, stdout, stderr}.
//
// Different sandbox images return data in different shapes:
//   - Wrapped (code_latest): {"exit_code":0,"stdout":"...","stderr":"..."}
//   - Raw (imgc-* / openclaw): the command's stdout returned verbatim
//
// We require strong signals to avoid mistaking user command JSON output
// for the wrapper:
//  1. Integer exit_code + at least one string stdout/stderr, OR
//  2. Both string stdout and stderr present.
func looksLikeWrappedShellPayload(data map[string]interface{}) bool {
	_, hasStdout := data["stdout"].(string)
	_, hasStderr := data["stderr"].(string)
	exitCodeVal, hasExitCode := data["exit_code"]
	hasIntExitCode := false
	if hasExitCode {
		if floatVal, ok := exitCodeVal.(float64); ok {
			hasIntExitCode = floatVal == float64(int(floatVal))
		}
	}
	if hasIntExitCode && (hasStdout || hasStderr) {
		return true
	}
	return hasStdout && hasStderr
}

// parseShellPayload parses a shell-tool MCP response payload.
// It detects the wrapped format conservatively; anything else is treated
// as raw command output returned verbatim.
func parseShellPayload(rawData string) parsedShellPayload {
	if rawData == "" {
		return parsedShellPayload{}
	}
	var dataMap map[string]interface{}
	if err := json.Unmarshal([]byte(rawData), &dataMap); err == nil {
		if looksLikeWrappedShellPayload(dataMap) {
			stdout, _ := dataMap["stdout"].(string)
			stderr, _ := dataMap["stderr"].(string)
			exitCode := 0
			if val, ok := dataMap["exit_code"].(float64); ok {
				exitCode = int(val)
			}
			traceID, _ := dataMap["traceId"].(string)
			return parsedShellPayload{
				stdout:   stdout,
				stderr:   stderr,
				exitCode: exitCode,
				traceID:  traceID,
			}
		}
	}
	// Raw stdout: return the entire payload verbatim.
	return parsedShellPayload{stdout: rawData}
}
